pkg/space/space.go

```diff
@@ -59,5 +61,18 @@ func NewSubSpace(spaceRequest *toolchainv1alpha1.SpaceRequest, parentSpace *tool
 
 // SubSpaceName generates a name for a subSpace based on parentSpace name and spacerequest UID.
 func SubSpaceName(parentSpace *toolchainv1alpha1.Space, spacerequest *toolchainv1alpha1.SpaceRequest) string {
-	return parentSpace.GetName() + string(spacerequest.UID[:5]) // we only get the first 5 chars from UID so we can keep the name within the length limits
+	parentSpaceName := parentSpace.GetName()
+
+	// if the parent space is itself a subspace, then we need to strip its UID
+	// suffix to prevent length limitations from kicking in
+	if parentSpace.Spec.ParentSpace != "" && len(parentSpaceName) > 6 {
+		// take off 6 characters to include the dash.  for example,
+		// "parentspace-12345" becomes "parentspace" when we strip the
+		// 5-character UID suffix and the "-"
+		parentSpaceName = string(parentSpaceName[:len(parentSpaceName)-6])
+	}
+
+	// only get the first 5 chars from the spacerequest's UID so we can keep
+	// the name within length limits.
+	return fmt.Sprintf("%v-%v", parentSpaceName, string(spacerequest.UID[:5]))
```

Comment:
how do you want to ensure that the already existing sub-spaces will be re-used after this change is merged?
1. there is a SpaceRequest CR in RHTAP instance
2. for the SpaceRequest CR, there is a sub-space `johny123` created
3. you merge the PR
4. the reconciler is kicked and it goes to this section https://github.com/codeready-toolchain/host-operator/blob/932b437277941dce8cc27b1e319931fd980d3cb0/controllers/spacerequest/spacerequest_controller.go#L183-L187
5. it uses this new format and expects `johny-123`, it doesn't find any, so it creates a new sub-space

the same would happen for the spaces whose parent has a name longer than 6

Reply:
good point ! But do we have any existent subSubSpace ? 

I mean the `parentSpaceName` will be shortened only if it's a subSpace it self, not only if the name is longer than 6:
`if parentSpace.Spec.ParentSpace != "" && len(parentSpaceName) > 6`

I thought right now we are using this feature only for 1 level of inheritance : `parentSpace --> subSpace` , and not yet `parentSpace --> subSpace --> subSubSpace` but maybe I'm wrong.

Reply:
In this particular case, could we list with a label selector for the correct parent space?
```go
subSpaceList := &toolchainv1alpha1.SpaceList{}
err := r.Client.List(ctx, subSpaceList,
    client.InNamespace(r.Namespace),
    client.MatchingLabels{
        toolchainv1alpha1.SpaceRequestLabelKey:          spaceRequest.GetName(),
        toolchainv1alpha1.SpaceRequestNamespaceLabelKey: spaceRequest.GetNamespace(),
        toolchainv1alpha1.ParentSpaceLabelKey:           parentSpace.GetName(),
    })
```

We'd have to handle the case where we find more than one matching space, but I can add a test for that too.

Reply:
> good point ! But do we have any existent subSubSpace ?

not of subSubSpace, but of subSpace yes. There is one more change in the logic of generating the name apart from the `if` statement that you mentioned above - originally, we didn't add the name between the prefix (parent name) and suffix (UID of the SpaceRequest CR). so the result was for example `parentspace1a2b3cd`. For better readability, we would like to add a hyphen between the prefix and suffix so its' `parentspace-1a2b3cd`, which however brings the challenge of keeping the backward compatibility.

The list for checking the existence should work. We had it there even before but we used it in a combination with generating the name nondeterministically. We modified it to the `Get` (as it is now) because we saw that the informer cache of the client wasn't sometimes updated quickly enough so the `List` returned an empty list even when one Space was already created. However, since we didn't use deterministic approach to generating the names, the we sometimes ended up with multiple Spaces per SpaceRequest CR.
Using the `List` should be fine, because it uses a fixed format for the names, so if the cache is not updated with the latest Space CR, then the next attempt to create the Space CR should fail with conflict, which totally OK. As a benefit, we don't have to rely on the naming format, but just on the labels set at Space CR.

Reply:
> There is one more change in the logic

My bad ! I missed that we are also adding the `-` between parentName and suffix 🤦‍♂️ . I don't remember why we haven't added that in the first place.

And indeed, if we want to change the naming convention of subSpaces as well, we could use `List` by labels while keeping the name deterministic.

Reply:
I've replaced most of the direct fetches with list calls in 563be3d (and added some tests in ff9eec2).  As far as I can tell, the only place that uses `SubSpaceName` anymore is some tests, and I'm half-tempted to refactor those tests to make those calls indirect, though I'll probably punt that off for a future PR if I do.